File: apps/vinto/src/app/services/bot-memory.ts
// services/bot-memory.ts
import { Card, Rank, Difficulty } from '../shapes';

/**
 * Memory entry for a single card observation
 */
export interface CardMemory {
  card: Card | null; // Actual card if known
  confidence: number; // 0-1, decays over time
  lastSeen: number; // Timestamp
  observations: number; // How many times seen
}

/**
 * Configuration for memory system based on difficulty
 */
export interface DifficultyMemoryConfig {
  memoryAccuracy: number; // 0-1: Chance to correctly remember a seen card
  memoryDecayRate: number; // How fast confidence decays over time (per second)
  maxMemorySize: number; // How many cards can be remembered
  forgetChance: number; // Chance to forget a card per turn
  observationRequired: number; // How many observations needed for high confidence
}

/**
 * Difficulty-based memory configurations
 */
export const DIFFICULTY_CONFIGS: Record<Difficulty, DifficultyMemoryConfig> = {
  easy: {
    memoryAccuracy: 0.4, // Often misremembers cards
    memoryDecayRate: 0.00015, // Fast decay (0.015% per second)
    maxMemorySize: 4, // Only remembers 4 cards
    forgetChance: 0.3, // 30% chance to forget per turn
    observationRequired: 3, // Needs to see multiple times
  },
  moderate: {
    memoryAccuracy: 0.75, // Usually correct
    memoryDecayRate: 0.00008, // Moderate decay
    maxMemorySize: 8, // Remembers 8 cards
    forgetChance: 0.1, // 10% chance to forget
    observationRequired: 2, // Needs 2 observations
  },
  hard: {
    memoryAccuracy: 0.95, // Almost perfect memory
    memoryDecayRate: 0.00002, // Slow decay
    maxMemorySize: 16, // Remembers almost everything
    forgetChance: 0.02, // Rarely forgets
    observationRequired: 1, // One observation enough
  },
};

/**
 * Bot memory system - tracks what the bot knows about cards in the game
 */
export class BotMemory {
  private config: DifficultyMemoryConfig;
  private botId: string;

  // Own cards knowledge (from setup peeks and actions)
  private ownCards: Map<number, CardMemory>;

  // Opponent cards knowledge (from peek actions)
  private opponentCards: Map<string, Map<number, CardMemory>>;

  // Deck tracking (cards seen in discard pile, etc.)
  private seenCards: Map<string, Card>; // cardId -> Card

  // Probabilistic beliefs about unknown cards
  private cardDistribution: Map<Rank, number>;

  constructor(botId: string, difficulty: Difficulty) {
    this.botId = botId;
    this.config = DIFFICULTY_CONFIGS[difficulty];
    this.ownCards = new Map();
    this.opponentCards = new Map();
    this.seenCards = new Map();
    this.cardDistribution = this.initializeCardDistribution();
  }

  /**
   * Initialize card distribution with standard deck composition
   */
  private initializeCardDistribution(): Map<Rank, number> {
    const distribution = new Map<Rank, number>();
    const ranks: Rank[] = [
      'A',
      '2',
      '3',
      '4',
      '5',
      '6',
      '7',
      '8',
      '9',
      '10',
      'J',
      'Q',
      'K',
    ];

    ranks.forEach((rank) => {
      // Standard deck: 4 of each rank
      distribution.set(rank, 4);
    });

    return distribution;
  }

  /**
   * Observe a card (from peeking, seeing in discard pile, etc.)
   */
  observeCard(card: Card, playerId: string, position: number): void {
    // Check if bot correctly remembers (based on memory accuracy)
    const isCorrect = Math.random() < this.config.memoryAccuracy;

    if (!isCorrect) {
      // Bot misremembers or fails to record
      console.log(
        `[BotMemory] ${this.botId} failed to remember card at ${playerId}[${position}]`
      );
      return;
    }

    // Update seen cards tracking
    this.seenCards.set(card.id, card);

    // Update card distribution
    const currentCount = this.cardDistribution.get(card.rank) || 0;
    if (currentCount > 0) {
      this.cardDistribution.set(card.rank, currentCount - 1);
    }

    // Store or update memory
    const targetMap =
      playerId === this.botId
        ? this.ownCards
        : this.getOrCreatePlayerMemory(playerId);
    const existing = targetMap.get(position);

    const newMemory: CardMemory = {
      card,
      confidence: existing ? Math.min(1.0, existing.confidence + 0.3) : 0.7,
      lastSeen: Date.now(),
      observations: existing ? existing.observations + 1 : 1,
    };

    targetMap.set(position, newMemory);

    console.log(
      `[BotMemory] ${this.botId} remembered ${
        card.rank
      } at ${playerId}[${position}] (confidence: ${newMemory.confidence.toFixed(
        2
      )})`
    );

    // Enforce memory size limit
    this.enforceMemoryLimit();
  }

  /**
   * Forget a card at a specific position (when card is removed or replaced)
   */
  forgetCard(playerId: string, position: number): void {
    if (playerId === this.botId) {
      const memory = this.ownCards.get(position);
      if (memory) {
        // Card was removed/replaced, update distribution
        const count = this.cardDistribution.get(memory.card!.rank) || 0;
        this.cardDistribution.set(memory.card!.rank, count + 1);
      }
      this.ownCards.delete(position);
    } else {
      const playerMemory = this.opponentCards.get(playerId);
      if (playerMemory) {
        const memory = playerMemory.get(position);
        if (memory) {
          const count = this.cardDistribution.get(memory.card!.rank) || 0;
          this.cardDistribution.set(memory.card!.rank, count + 1);
        }
        playerMemory.delete(position);
      }
    }
  }

  /**
   * Get memory for a specific card position
   */
  getCardMemory(playerId: string, position: number): CardMemory | undefined {
    if (playerId === this.botId) {
      return this.ownCards.get(position);
    } else {
      return this.opponentCards.get(playerId)?.get(position);
    }
  }

  /**
   * Get all known cards for a player
   */
  getPlayerMemory(playerId: string): Map<number, CardMemory> {
    if (playerId === this.botId) {
      return new Map(this.ownCards);
    } else {
      return new Map(this.opponentCards.get(playerId) || new Map());
    }
  }

  /**
   * Get confidence about a card at a position (0-1)
   */
  getConfidence(playerId: string, position: number): number {
    const memory = this.getCardMemory(playerId, position);
    return memory ? memory.confidence : 0;
  }

  /**
   * Decay all memories over time
   */
  decayMemory(): void {
    const now = Date.now();
    const decayRate = this.config.memoryDecayRate;

    // Decay own cards
    this.decayMemoryMap(this.ownCards, now, decayRate);

    // Decay opponent cards
    for (const [, playerMemory] of this.opponentCards) {
      this.decayMemoryMap(playerMemory, now, decayRate);
    }
  }

  /**
   * Helper to decay a memory map
   */
  private decayMemoryMap(
    memoryMap: Map<number, CardMemory>,
    now: number,
    decayRate: number
  ): void {
    const toDelete: number[] = [];

    for (const [position, memory] of memoryMap) {
      const timeSinceLastSeen = now - memory.lastSeen;
      const decay = Math.exp(-decayRate * timeSinceLastSeen);

      memory.confidence *= decay;

      // Forget if confidence too low
      if (memory.confidence < 0.1) {
        toDelete.push(position);
      }
    }

    // Remove forgotten cards
    toDelete.forEach((pos) => memoryMap.delete(pos));
  }

  /**
   * Random forgetting at turn boundaries
   */
  processTurnBoundary(): void {
    // Random chance to forget cards
    this.randomForget(this.ownCards);

    for (const [, playerMemory] of this.opponentCards) {
      this.randomForget(playerMemory);
    }

    // Decay all memories
    this.decayMemory();
  }

  /**
   * Randomly forget some cards
   */
  private randomForget(memoryMap: Map<number, CardMemory>): void {
    const toDelete: number[] = [];

    for (const [position] of memoryMap) {
      if (Math.random() < this.config.forgetChance) {
        toDelete.push(position);
      }
    }

    toDelete.forEach((pos) => {
      const memory = memoryMap.get(pos);
      if (memory) {
        // Return card to distribution
        const count = this.cardDistribution.get(memory.card!.rank) || 0;
        this.cardDistribution.set(memory.card!.rank, count + 1);
      }
      memoryMap.delete(pos);
    });

    if (toDelete.length > 0) {
      console.log(`[BotMemory] ${this.botId} forgot ${toDelete.length} cards`);
    }
  }

  /**
   * Enforce memory size limit - forget lowest confidence cards
   */
  private enforceMemoryLimit(): void {
    const allMemories: Array<{
      playerId: string;
      position: number;
      confidence: number;
      map: Map<number, CardMemory>;
    }> = [];

    // Collect all memories
    for (const [position, memory] of this.ownCards) {
      allMemories.push({
        playerId: this.botId,
        position,
        confidence: memory.confidence,
        map: this.ownCards,
      });
    }

    for (const [playerId, playerMemory] of this.opponentCards) {
      for (const [position, memory] of playerMemory) {
        allMemories.push({
          playerId,
          position,
          confidence: memory.confidence,
          map: playerMemory,
        });
      }
    }

    // Sort by confidence (ascending)
    allMemories.sort((a, b) => a.confidence - b.confidence);

    // Forget lowest confidence cards if over limit
    const toForget = allMemories.length - this.config.maxMemorySize;
    if (toForget > 0) {
      for (let i = 0; i < toForget; i++) {
        const item = allMemories[i];
        const memory = item.map.get(item.position);
        if (memory) {
          // Return to distribution
          const count = this.cardDistribution.get(memory.card!.rank) || 0;
          this.cardDistribution.set(memory.card!.rank, count + 1);
        }
        item.map.delete(item.position);
      }
      console.log(
        `[BotMemory] ${this.botId} enforced memory limit, forgot ${toForget} cards`
      );
    }
  }

  /**
   * Get or create memory map for a player
   */
  private getOrCreatePlayerMemory(playerId: string): Map<number, CardMemory> {
    if (!this.opponentCards.has(playerId)) {
      this.opponentCards.set(playerId, new Map());
    }
    return this.opponentCards.get(playerId)!;
  }

  /**
   * Get probability distribution for unknown cards
   */
  getCardDistribution(): Map<Rank, number> {
    return new Map(this.cardDistribution);
  }

  /**
   * Sample a card from the current distribution
   */
  sampleCardFromDistribution(): Rank | null {
    const available: Rank[] = [];

    for (const [rank, count] of this.cardDistribution) {
      for (let i = 0; i < count; i++) {
        available.push(rank);
      }
    }

    if (available.length === 0) return null;

    return available[Math.floor(Math.random() * available.length)];
  }

  /**
   * Get total number of cards in memory
   */
  getMemorySize(): number {
    let size = this.ownCards.size;

    for (const [, playerMemory] of this.opponentCards) {
      size += playerMemory.size;
    }

    return size;
  }

  /**
   * Get memory statistics for debugging
   */
  getStats() {
    return {
      ownCardsKnown: this.ownCards.size,
      opponentCardsKnown: Array.from(this.opponentCards.values()).reduce(
        (sum, map) => sum + map.size,
        0
      ),
      totalMemoryUsed: this.getMemorySize(),
      maxMemorySize: this.config.maxMemorySize,
      seenCardsCount: this.seenCards.size,
    };
  }

  /**
   * Clear all memory (for testing/reset)
   */
  clear(): void {
    this.ownCards.clear();
    this.opponentCards.clear();
    this.seenCards.clear();
    this.cardDistribution = this.initializeCardDistribution();
  }
}

File: kmp/shared/bot/src/commonMain/kotlin/game/vinto/bot/BotMemory.kt
package game.vinto.bot

import game.vinto.shapes.ALL_RANKS
import game.vinto.shapes.Card
import game.vinto.shapes.Difficulty
import game.vinto.shapes.Rank
import kotlin.math.exp
import kotlin.math.min
import kotlin.random.Random

/**
 * What a bot remembers, and how imperfectly.
 *
 * Difficulty is not implemented by making the bot think worse — it thinks the same at every
 * level — but by how reliably it *remembers*. An easy bot misremembers cards, forgets them
 * quickly and can only hold four at a time; a hard bot remembers everything perfectly. That
 * is a far better model of a weak player than deliberately choosing bad moves.
 *
 * Ported from `packages/bot/src/lib/bot-memory.ts`, with two things injected that were
 * ambient there:
 *
 *  - **[Random]**, replacing `Math.random`. Design D4 requires it, and it is what makes a
 *    bot's play reproducible from a seed. It is also the direct cause of the two known-flaky
 *    bot tests in the TypeScript suite.
 *  - **the clock**, replacing `Date.now()`. Memory decays with elapsed time, so a bot reading
 *    a wall clock plays differently depending on how long a turn took — which would make a
 *    recorded game unreplayable. [ticks] is advanced by the caller at turn boundaries.
 */
data class DifficultyMemoryConfig(
    /** Chance of correctly recording a card that was seen. */
    val memoryAccuracy: Double,
    /** How fast confidence decays per tick. */
    val memoryDecayRate: Double,
    val maxMemorySize: Int,
    /** Chance of dropping any given memory at a turn boundary. */
    val forgetChance: Double,
    val observationRequired: Int,
)

val DIFFICULTY_CONFIGS: Map<Difficulty, DifficultyMemoryConfig> = mapOf(
    Difficulty.EASY to DifficultyMemoryConfig(
        memoryAccuracy = 0.4,
        memoryDecayRate = 0.00015,
        maxMemorySize = 4,
        forgetChance = 0.3,
        observationRequired = 3,
    ),
    Difficulty.MODERATE to DifficultyMemoryConfig(
        memoryAccuracy = 0.75,
        memoryDecayRate = 0.00008,
        maxMemorySize = 8,
        forgetChance = 0.1,
        observationRequired = 2,
    ),
    // Perfect and non-decaying, which is also what makes hard-difficulty tests deterministic.
    Difficulty.HARD to DifficultyMemoryConfig(
        memoryAccuracy = 1.0,
        memoryDecayRate = 0.0,
        maxMemorySize = 100,
        forgetChance = 0.0,
        observationRequired = 1,
    ),
)

data class CardMemory(
    val card: Card,
    /** 0-1, decaying with time and rising with repeated sightings. */
    val confidence: Double,
    val lastSeen: Long,
    val observations: Int,
)

private const val INITIAL_CONFIDENCE = 0.7
private const val REOBSERVATION_BOOST = 0.3
private const val FORGET_BELOW_CONFIDENCE = 0.1
private const val COPIES_PER_RANK = 4
private const val JOKER_COUNT = 2

/** Fallback when nothing is known about what is left; roughly a mid-deck card. */
private const val NEUTRAL_AVERAGE_CARD_VALUE = 6.0

/** A memory below this confidence is a hunch, not a fact, and is estimated rather than trusted. */


class BotMemory(
    private val botId: String,
    difficulty: Difficulty,
    private val random: Random = Random.Default,
) {
    private val config = DIFFICULTY_CONFIGS.getValue(difficulty)

    private val ownCards = mutableMapOf<Int, CardMemory>()
    private val opponentCards = mutableMapOf<String, MutableMap<Int, CardMemory>>()

    /**
     * Ranks lying in plain sight — the discard pile and the card in play — as last synced
     * from the table by [syncVisibleCards]. Public facts, not memories: they are never
     * misremembered, never decay, and shrink on their own when a reshuffle folds the pile
     * back into the deck.
     */
    private var visibleRanks: List<Rank> = emptyList()

    /**
     * The bot's own sense of elapsed time, advanced by [processTurnBoundary] rather than read
     * from a clock. One tick per turn keeps decay meaningful and the whole thing replayable.
     */
    private var ticks: Long = 0

    private fun initialDistribution(): MutableMap<Rank, Int> =
        ALL_RANKS.associateWith { rank ->
            if (rank == Rank.JOKER) JOKER_COUNT else COPIES_PER_RANK
        }.toMutableMap()

    /**
     * Records a card the bot has seen — and, depending on difficulty, sometimes fails to.
     * A failed observation is silent, exactly as it would be for a person.
     */
    fun observeCard(card: Card, playerId: String, position: Int) {
        if (random.nextDouble() >= config.memoryAccuracy) return

        val target = memoryFor(playerId)
        val existing = target[position]
        target[position] = CardMemory(
            card = card,
            confidence = existing?.let { min(1.0, it.confidence + REOBSERVATION_BOOST) }
                ?: INITIAL_CONFIDENCE,
            lastSeen = ticks,
            observations = (existing?.observations ?: 0) + 1,
        )

        enforceMemoryLimit()
    }

    /** A card that moved or left; the memory is wrong now. The pool recovers on its own. */
    fun forgetCard(playerId: String, position: Int) {
        memoryFor(playerId).remove(position)
    }

    /** The table's public cards, re-synced before each decision. See [visibleRanks]. */
    fun syncVisibleCards(ranks: List<Rank>) {
        visibleRanks = ranks.toList()
    }

    fun getCardMemory(playerId: String, position: Int): CardMemory? = memoryFor(playerId)[position]

    fun getPlayerMemory(playerId: String): Map<Int, CardMemory> = memoryFor(playerId).toMap()

    /**
     * What this bot would declare about its own hand: the positions it trusts, as ranks.
     * On lower difficulties observation can record the wrong card or nothing at all, so a
     * declaration built from this can be honestly wrong — which is the point.
     */
    fun believedOwnCards(): Map<Int, Rank> = ownCards
        .filterValues { it.confidence > TRUSTED_CONFIDENCE }
        .mapValues { (_, memory) -> memory.card.rank }

    fun getConfidence(playerId: String, position: Int): Double =
        getCardMemory(playerId, position)?.confidence ?: 0.0

    /**
     * What is still unaccounted for: a full deck minus everything this bot can point to — the
     * public cards on the table and every memory it trusts, its own hand's and its
     * opponents'.
     *
     * *Derived*, never mutated. The old bookkeeping decremented on observation and credited
     * back on forgetting, and leaked on every path those two did not cover: an overwritten
     * memory, the same card re-seen at a new position, a discard nobody observed, a
     * reshuffle, a new deal. Recomputing from what is currently known makes each of those a
     * non-event.
     */
    fun getCardDistribution(): Map<Rank, Int> {
        val remaining = initialDistribution()
        fun consume(rank: Rank) {
            remaining[rank]?.let { count -> if (count > 0) remaining[rank] = count - 1 }
        }

        visibleRanks.forEach(::consume)
        ownCards.values.forEach { if (it.confidence > TRUSTED_CONFIDENCE) consume(it.card.rank) }
        opponentCards.values.forEach { map ->
            map.values.forEach { if (it.confidence > TRUSTED_CONFIDENCE) consume(it.card.rank) }
        }
        return remaining
    }

    fun getMemorySize(): Int = ownCards.size + opponentCards.values.sumOf { it.size }

    /** Advances the bot's clock, then forgets and decays. Call once per turn. */
    fun processTurnBoundary() {
        ticks++
        randomForget(ownCards)
        opponentCards.values.forEach { randomForget(it) }
        decayMemory()
    }

    fun decayMemory() {
        decayMemoryMap(ownCards)
        opponentCards.values.forEach { decayMemoryMap(it) }
    }

    /** Draws a plausible rank for an unseen card, weighted by what is still unaccounted for. */
    fun sampleCardFromDistribution(): Rank? {
        val available = getCardDistribution().entries.flatMap { (rank, count) -> List(count) { rank } }
        if (available.isEmpty()) return null
        return available[random.nextInt(available.size)]
    }

    fun clear() {
        ownCards.clear()
        opponentCards.clear()
        visibleRanks = emptyList()
        ticks = 0
    }

    private fun memoryFor(playerId: String): MutableMap<Int, CardMemory> =
        if (playerId == botId) ownCards else opponentCards.getOrPut(playerId) { mutableMapOf() }

    private fun decayMemoryMap(memoryMap: MutableMap<Int, CardMemory>) {
        if (config.memoryDecayRate == 0.0) return

        val forgotten = mutableListOf<Int>()
        for ((position, memory) in memoryMap) {
            val elapsed = (ticks - memory.lastSeen).toDouble()
            val decayed = memory.confidence * exp(-config.memoryDecayRate * elapsed)
            if (decayed < FORGET_BELOW_CONFIDENCE) {
                forgotten += position
            } else {
                memoryMap[position] = memory.copy(confidence = decayed)
            }
        }
        forgotten.forEach { memoryMap.remove(it) }
    }

    private fun randomForget(memoryMap: MutableMap<Int, CardMemory>) {
        if (config.forgetChance == 0.0) return

        memoryMap.keys.filter { random.nextDouble() < config.forgetChance }
            .forEach { memoryMap.remove(it) }
    }

    /**
     * Over the limit, the least-certain memories go first — which is how a person forgets.
     *
     * The limit is a cap on *opponent* tracking only. The bot's own row is the thing it
     * rehearses every turn and is never crowded out by watching the table — when it was,
     * every sighting of an opponent's card evicted one of the bot's own, `believedOwnCards`
     * went empty for any bot that had seen a few opponents, and no bot on a limited
     * difficulty could ever believe its full hand — which meant it could never call Vinto,
     * and games stopped ending. Own-hand fallibility comes from [DifficultyMemoryConfig]'s
     * accuracy and forget chance instead, which is the honest kind: gaps, not amnesia.
     */
    private fun enforceMemoryLimit() {
        val over = opponentCards.values.sumOf { it.size } - config.maxMemorySize
        if (over <= 0) return

        val all = opponentCards.values.flatMap { map ->
            map.map { (position, memory) -> Triple(map, position, memory) }
        }

        all.sortedBy { it.third.confidence }
            .take(over)
            .forEach { (map, position, _) -> map.remove(position) }
    }
}

/**
 * The bot's estimate of a player's total, mixing what it remembers with what it can infer
 * about the rest. Only memories it is more than half confident in are trusted as fact.
 *
 * Ported from `packages/bot/src/lib/mcts-score-estimator.ts`.
 */
fun estimatePlayerScore(handSize: Int, botMemory: BotMemory, playerId: String): Double {
    val known = botMemory.getPlayerMemory(playerId)

    var knownScore = 0.0
    var unknownCount = 0
    for (position in 0 until handSize) {
        val memory = known[position]
        if (memory != null && memory.confidence > TRUSTED_CONFIDENCE) knownScore += memory.card.value
        else unknownCount++
    }

    return knownScore + unknownCount * averageRemainingCardValue(botMemory)
}

/** Average value of everything still unaccounted for. */
fun averageRemainingCardValue(botMemory: BotMemory): Double {
    var totalValue = 0.0
    var totalCount = 0

    for ((rank, count) in botMemory.getCardDistribution()) {
        if (count > 0) {
            totalValue += game.vinto.shapes.getCardValue(rank) * count
            totalCount += count
        }
    }

    return if (totalCount == 0) NEUTRAL_AVERAGE_CARD_VALUE else totalValue / totalCount
}
